mlir/include/mlir/Dialect/Flux/IR/FluxOps.td

```diff
@@ -0,0 +1,411 @@
+        - `register_size`: Total size of the register
+        - `register_index`: Index within the register for this measurement
+
+        Example (simple measurement):
+        ```mlir
+        %q_out, %result = flux.measure %q_in : !flux.qubit
+        ```
+
+        Example (measurement with output recording):
+        ```mlir
+        %q_out, %result = flux.measure("c", 2, 0) %q_in : !flux.qubit
+        ```
+    }];
+
+    let arguments = (ins Arg<QubitType, "the qubit to measure", [MemRead]>:$qubit_in,
+                     OptionalAttr<StrAttr>:$register_name,
+                     OptionalAttr<ConfinedAttr<I64Attr, [IntPositive]>>:$register_size,
+                     OptionalAttr<ConfinedAttr<I64Attr, [IntNonNegative]>>:$register_index);
+    let results = (outs QubitType:$qubit_out, I1:$result);
+    let assemblyFormat = [{
+        (`(` $register_name^ `,` $register_size `,` $register_index `)`)?
+        $qubit_in `:` type($qubit_in) attr-dict
+    }];
+
+    let builders = [
+        OpBuilder<(ins "Value":$qubit_in), [{
+            build($_builder, $_state, QubitType::get($_builder.getContext()), $_builder.getI1Type(),
+                  qubit_in, nullptr, nullptr, nullptr);
+        }]>
+    ];
+
+    let hasVerifier = 1;
+}
+
+def ResetOp : FluxOp<"reset", [Idempotent, SameOperandsAndResultType]> {
+    let summary = "Reset a qubit to |0⟩ state";
+    let description = [{
+        Resets a qubit to the |0⟩ state, regardless of its current state,
+        and returns the reset qubit.
+
+        Example:
+        ```mlir
+        %q_out = flux.reset %q_in : !flux.qubit -> !flux.qubit
+        ```
+    }];
+
+    let arguments = (ins Arg<QubitType, "the qubit to reset", [MemRead]>:$qubit_in);
+    let results = (outs QubitType:$qubit_out);
+    let assemblyFormat = "$qubit_in attr-dict `:` type($qubit_in) `->` type($qubit_out)";
+    let hasCanonicalizer = 1;
+}
+
+//===----------------------------------------------------------------------===//
+// Unitary Operations
+//===----------------------------------------------------------------------===//
+
+def XOp : FluxOp<"x", traits = [Flux_UnitaryOpInterface]> {
+    let arguments = (ins QubitType:$qubit_in);
+    let results = (outs QubitType:$qubit_out);
+    let assemblyFormat = "$qubit_in attr-dict `:` type($qubit_in) `->` type($qubit_out)";
+
+    let extraClassDeclaration = [{
+        size_t getNumQubits() { return getNumTargets() + getNumControls(); }
+        static size_t getNumTargets() { return 1; }
+        size_t getNumControls() { return getNumPosControls() + getNumNegControls(); }
+        static size_t getNumPosControls() { return 0; }
+        static size_t getNumNegControls() { return 0; }
+        Value getQubit(size_t i);
+        Value getTarget(size_t i);
+        static Value getPosControl(size_t i);
+        static Value getNegControl(size_t i);
+        Value getInput(size_t i);
+        Value getOutput(size_t i);
```

Comment:
Hm. I see a slight overlap in the getQubit and getInput methods here.
They effectively do the same thing right?
Might it make sense to eliminate the getQubit method for the flux interface?

A similar question arises for the getter of the controls and targets. Shouldn't these be two methods each for the inputs and outputs?